Support kernel_info request on the control channel (#82)
jupyter/enhancement-proposals: 80-kernel-info/kernel-info.md

> # Send kernel\_info request on the control channel

Reviewer: ```suggestion
# Support kernel\_info request on the control channel
```

## Problem

> When connecting a new websocket to an existing kernel via the Jupyter server, if the kernel execution is
> stopped on a breakpoint, the messages sent over the websocket get no reply. This is because when

Reviewer 2: ```suggestion
When connecting a new websocket to an existing kernel via the Jupyter server, if the kernel execution is
busy (e.g. long running cell) or stopped (e.g. on a breakpoint), the messages sent over the websocket get no reply. This is because when
```

establishing a new websocket connection, the Jupyter server will send `kernel_info` requests to the kernel
and will prevent sending any other request until it receives a `kernel_info` reply. Since the `kernel_info`
request is sent on the shell channel and the kernel execution is stopped, it cannot reply to that request.

## Proposed enhancement

We propose to state in the Jupyter Messaging Protocol that the `kernel_info` request can be sent on both the shell and the control channels. Although both channels supports the `kernel_info` message, clients are encouraged to send it on the control channel as it will always be able to handle it, while the shell channel may be stopped.

### Impact on existing implementations

> This JEP impacts kernels since it requires them to support receiving 'kernel\_info\_request' on the dontrol channel in addition to receiving them on the shell channel.

Reviewer: ```suggestion
This JEP impacts kernels since it requires them to support receiving 'kernel\_info\_request' on the control channel in addition to receiving them on the shell channel.
```

> It also has an impact on  the Jupyter Server, which must be updated accordingly.

Reviewer 2: ```suggestion
It also has an impact on the Jupyter Server. For example, the reference implementation of Jupyter Server will attempt to send a a `kernel_info` request on both channels and listen for a response from _either_ channel. Any response informs the UI that the kernel is connected. 
```

# Send kernel\_info request on the control channel

## Problem

When connecting a new websocket to an existing kernel via the Jupyter server, if the kernel execution is
stopped on a breakpoint, the messages sent over the websocket get no reply. This is because when
establibshing a new websocket connection, the Jupyter server will send `kernel_info` requests to the kernel
and will prevent sending any other request until it receives a `kernel_info` reply. Since the `kernel_info`

> request is sent on the shell channel and the kernel executino is stopped, it cannot reply to that request.

Reviewer 3: ```suggestion
request is sent on the shell channel and the kernel execution is stopped, it cannot reply to that request.
```

## Proposed enhancement

> We propose to state in the Jupyter Messaging Protocol that the `kernel_info` request muter st be sent on the

Reviewer 3: ```suggestion
We propose to state in the Jupyter Messaging Protocol that the `kernel_info` request must be sent on the
```

control channel exclusively.

### Impact on existing implementations

There should not be any impact on the existing kernels, since the current specification states that any message
that can be sent on the shell channel could be sent on the control channel.

The only impact of this change (beyond the protocol itself) is on the Jupyter Server, which must be updated
accordingly.

## Relevant Resources (GitHub repositories, Issues, PRs)

### GitHub repositories

> - [Jupyter server](https://github.com/jupyter-server/jupyter_server): the backend to Jupyter web applications

Reviewer 3: Maybe add jupyverse?
```suggestion
- [Jupyter server](https://github.com/jupyter-server/jupyter_server): the official backend to Jupyter web applications
- [Jupyverse](https://github.com/jupyter-server/jupyverse): a FastAPI-based backend to Jupyter web applications
```

# Send kernel\_info request on the control channel

## Problem

When connecting a new websocket to an existing kernel via the Jupyter server, if the kernel execution is
stopped on a breakpoint, the messages sent over the websocket get no reply. This is because when

> establibshing a new websocket connection, the Jupyter server will send `kernel_info` requests to the kernel

Reviewer: ```suggestion
establishing a new websocket connection, the Jupyter server will send `kernel_info` requests to the kernel
```
